File: backend/src/EPMS.Application/DTOs/Common/PagedResult.cs
namespace EPMS.Application.DTOs.Common;

public class PagedResult<T>
{
    public IReadOnlyList<T> Items { get; init; } = Array.Empty<T>();

    public int Page { get; init; }

    public int PageSize { get; init; }

    public int TotalItems { get; init; }

    public int TotalPages => PageSize == 0 ? 0 : (int)Math.Ceiling(TotalItems / (double)PageSize);

    public bool HasNextPage => Page < TotalPages;

    public bool HasPreviousPage => Page > 1;

    public static PagedResult<T> Create(IReadOnlyList<T> items, int totalItems, int page, int pageSize) =>
        new()
        {
            Items = items,
            TotalItems = totalItems,
            Page = page,
            PageSize = pageSize
        };
}

File: frontend/src/shared/api/paged-result.ts
// Mencerminkan EPMS.Application.DTOs.Common.PagedResult<T> dari backend.
// Dipakai bersama oleh project-api.ts, user-api.ts, dll — ditaruh di shared
// supaya tidak ada feature yang mengimpor tipe dari feature lain secara langsung.
export interface PagedResult<T> {
  items: T[];
  page: number;
  pageSize: number;
  totalItems: number;
  totalPages: number;
  hasNextPage: boolean;
  hasPreviousPage: boolean;
}
